### PR TITLE
Feature/1603 ens offchain integration

### PR DESCRIPTION
**Pull request by andrii-kl**
### What does this PR do?
Add support of new functions of ENS. eip-3668 | EIP 1577

### Why is it needed?
To work with new ENS functions.

### CODE REVIEW

**Comment by makoto** [comment]
Hi @andrii-kl thanks for the integration PR and sorry for not being able to respond quickly.
I have a couple of comments

1. Does it have any usage example in doc or in the test?
2. Could you test with real data `1.offchainexample.eth` as described [in our doc](https://docs.ens.domains/dapp-developer-guide/ens-l2-offchain#ethersjs) and see if it returns the record as expected?

Thanks.

Makoto

**Comment by makoto** [comment]
FYI. Added some test casce snenario for web3.py which you can refer to as well https://github.com/ethereum/web3.py/issues/2406

**Comment by andrii-kl** [comment]
@makoto  
I have change a bit my realisation according to example from 
https://docs.ens.domains/ens-improvement-proposals/ensip-10-wildcard-resolution
but I have an error when I try to call resolver.resolve(dnsEncodedName, calldata) 
Invalid ENS name provided: 0x3b3b57de1c9fb...........

I'm not sure how to check if my CallData is correct.

**Comment by andrii-kl** [comment]
@makoto  I try to test with the 1.offchainexample.eth, I've got the same response when in the etherscan interface.
Screen Shot 2022-04-29 at 12 47 18 PM

Could you provide right execution parameters for this method? (1.offchainexample.eth)

File: core/src/main/java/org/web3j/ens/NameHash.java [hunk 1]
```diff
@@ -69,4 +71,36 @@ public static String normalise(String ensName) {
...
+            byte[] bytes = toUtf8Bytes("_" + normalise(part));
+            if (bytes == null) {
+                break;
+            }
+            bytes[0] = (byte) (bytes.length - 1);
+
+            outputStream.write(bytes);
+        }
+
+        return Numeric.toHexString(outputStream.toByteArray()) + "00";
```
Comment (gtebrean) [inline core/src/main/java/org/web3j/ens/NameHash.java:104]: just to know, why is necessary to append 00?

File: core/src/main/java/org/web3j/ens/contracts/generated/ENSRegistryWithFallback.java [hunk 2]
```diff
@@ -0,0 +1,439 @@
...
+import org.web3j.tx.gas.ContractGasProvider;
+
+/**
+ * <p>Auto generated code.
+ * <p><strong>Do not modify!</strong>
+ * <p>Please use the <a href="https://docs.web3j.io/command_line.html">web3j command line tools</a>,
+ * or the org.web3j.codegen.SolidityFunctionWrapperGenerator in the 
+ * <a href="https://github.com/web3j/web3j/tree/master/codegen">codegen module</a> to update.
+ *
+ * <p>Generated with web3j version 1.4.1.
```
Comment (gtebrean) [inline core/src/main/java/org/web3j/ens/contracts/generated/ENSRegistryWithFallback.java:37]: shouldn't this be generated with a newer version?

File: core/src/main/java/org/web3j/ens/EnsResolver.java [hunk 3]
```diff
@@ -78,40 +105,193 @@ protected PublicResolver obtainPublicResolver(String ensName) {
...
+                        resolver.supportsInterface(EnsUtils.ENSIP_10_INTERFACE_ID).send();
+
+                byte[] nameHash = NameHash.nameHashAsBytes(ensName);
+                String dnsEncoded = NameHash.dnsEncode(ensName);
+
+                String contractAddress;
+
+                if (supportWildcard) {
+                    String callData = resolver.addr(nameHash).encodeFunctionCall();
+                    String to = resolver.getContractAddress();
```
Comment (gtebrean) [inline core/src/main/java/org/web3j/ens/EnsResolver.java:160]: assigned but no further used🤔  
Didn't you want to assigned this to contractAddress above?

**Review by gtebrean — COMMENTED** [review]
looks good for me, only few questions

**Comment by makoto** [comment]
> @makoto Same response in the etherscan too. (ensName: 1.offchainexample.eth) Screen Shot 2022-04-29 at 12 47 18 PM
> 
> Could you provide right execution parameters for this method? (1.offchainexample.eth)

That should show custom error type with gateway URL like the below as long as you put two bytecode

````
     Error: call revert exception [ See: https://links.ethers.org/v5-errors-CALL_EXCEPTION ] (method="resolve(bytes,bytes)", errorArgs=["0x5FbDB2315678afecb367f032d93F642f64180aa3",["http://localhost:8080/"],"0x9061b92300000000000000000000000000000000000000000000000000000000000000400000000000000000000000000000000000000000000000000000000000000080000000000000000000000000000000000000000000000000000000000000001701310f6f6666636861696e6578616d706c6503657468000000000000000000000000000000000000000000000000000000000000000000000000000000000000","0xf4d4d2f8","0x9061b92300000000000000000000000000000000000000000000000000000000000000400000000000000000000000000000000000000000000000000000000000000080000000000000000000000000000000000000000000000000000000000000001701310f6f6666636861696e6578616d706c6503657468000000000000000000000000000000000000000000000000000000000000000000000000000000000000"], errorName="OffchainLookup", errorSignature="OffchainLookup(address,string[],bytes,bytes4,bytes)", reason=null, code=CALL_EXCEPTION, version=abi/5.6.0)
```

For some reason, it's not showing as you demonstrated. Let me check with etherscan team if they do support showing custom error type

File: core/src/main/java/org/web3j/ens/NameHash.java (hunk 1 above)
Comment (andrii-kl) [reply core/src/main/java/org/web3j/ens/NameHash.java:104]: I think it is used as a marker of the end.

**Comment by andrii-kl** [comment]
> > @makoto Same response in the etherscan too. (ensName: 1.offchainexample.eth) Screen Shot 2022-04-29 at 12 47 18 PM
> > Could you provide right execution parameters for this method? (1.offchainexample.eth)
> 
> That should show custom error type with gateway URL like the below as long as you put two bytecode
> 
> ```
>      Error: call revert exception [ See: https://links.ethers.org/v5-errors-CALL_EXCEPTION ] (method="resolve(bytes,bytes)", errorArgs=["0x5FbDB2315678afecb367f032d93F642f64180aa3",["http://localhost:8080/"],"0x9061b92300000000000000000000000000000000000000000000000000000000000000400000000000000000000000000000000000000000000000000000000000000080000000000000000000000000000000000000000000000000000000000000001701310f6f6666636861696e6578616d706c6503657468000000000000000000000000000000000000000000000000000000000000000000000000000000000000","0xf4d4d2f8","0x9061b92300000000000000000000000000000000000000000000000000000000000000400000000000000000000000000000000000000000000000000000000000000080000000000000000000000000000000000000000000000000000000000000001701310f6f6666636861696e6578616d706c6503657468000000000000000000000000000000000000000000000000000000000000000000000000000000000000"], errorName="OffchainLookup", errorSignature="OffchainLookup(address,string[],bytes,bytes4,bytes)", reason=null, code=CALL_EXCEPTION, version=abi/5.6.0)
> ```
> 
> For some reason, it's not showing as you demonstrated. Let me check with etherscan team if they do support showing custom error type
> ```

Yes, it would be great to add support on the etherscan. 

On our side, everything is fine, we have dealt with everything, thank you.

**Review by gtebrean — APPROVED** [review]

**Pull request merged**